pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java

```diff
@@ -481,6 +481,17 @@ static void validatePartialUpsertStrategies(TableConfig tableConfig, Schema sche
       UpsertConfig.Strategy columnStrategy = entry.getValue();
       Preconditions.checkState(!primaryKeyColumns.contains(column), "Merger cannot be applied to primary key columns");
 
+      if (upsertConfig.getComparisonColumn() != null) {
+        Preconditions.checkState(!upsertConfig.getComparisonColumn().equals(column),
+            "Merger cannot be applied to comparison column");
+      }
+
+      if (upsertConfig.getComparisonColumn() == null
+          && tableConfig.getValidationConfig().getTimeColumnName() != null) {
+        Preconditions.checkState(!tableConfig.getValidationConfig().getTimeColumnName().equals(column),
+            "Merger cannot be applied to time column");
+      }
```

Comment:
Time column must exist for upsert table (real-time table), so this check can be simplified:
```suggestion
      if (upsertConfig.getComparisonColumn() != null) {
        Preconditions.checkState(!upsertConfig.getComparisonColumn().equals(column),
            "Merger cannot be applied to comparison column");
      } else {
        Preconditions.checkState(!tableConfig.getValidationConfig().getTimeColumnName().equals(column),
            "Merger cannot be applied to time column");
      }
```

Reply:
A minor optimization on top of this is to pre-compute the comparison column before getting into the loop:
```
String comparisonColumn = upsertConfig.getComparisonColumn();
if (comparisonColumn == null) {
  comparisonColumn = tableConfig.getValidationConfig().getTimeColumnName();
}
```
Then just check if the column is comparison column:
```
Preconditions.checkState(!comparisonColumn.equals(column), "Merger cannot be applied to comparison column");
```

Reply:
I didn't do the optimisation because I want to make it explicit to the user what column is causing the issue. User might be confused that although he has not defined any comparison column, it is still saying invalid config.